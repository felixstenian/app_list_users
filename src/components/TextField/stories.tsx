import { Meta, Story } from '@storybook/react'
import TextField, { TextFieldProps } from '.'

export default {
  title: 'TextField',
  component: TextField,
  args: {}
} as Meta

export const Default: Story<TextFieldProps> = (args) => <TextField {...args} />
