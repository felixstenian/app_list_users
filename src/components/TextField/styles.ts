import styled, { css } from 'styled-components'
import { layout, space } from 'styled-system'

import { TextFieldProps } from '.'

type IconPositionProps = Pick<TextFieldProps, 'iconPosition'>

export const Wrapper = styled.div<TextFieldProps>`
  ${space};
  ${layout};
`

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.black};
    cursor: pointer;
  `}
`

export const InputWraper = styled.div`
  display: flex;
  align-items: center;
  border-radius: 4rem;
  border: 0.2rem solid;

  ${({ theme }) => css`
    background: ${theme.colors.lightGray};
    padding: ${theme.spacings.xxsmall};
    border-color: ${theme.colors.lightGray};

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.gray};
    }
  `}
`

export const Icon = styled.div<IconPositionProps>`
  ${({ theme, iconPosition }) => css`
    display: flex;
    color: ${theme.colors.gray};
    order: ${iconPosition === 'right' ? 1 : 0};
    & > svg {
      width: 2.2rem;
      height: 100%;
    }
  `}
`

export const Input = styled.input<IconPositionProps>`
  ${({ theme, iconPosition }) => css`
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} 0;
    padding-${iconPosition}: ${theme.spacings.xsmall};
    background: transparent;
    border: 0;
    outline: none;
    width: ${iconPosition === 'right' ? `calc(100% - 2.2rem)` : `100%`};
    &:-webkit-autofill {
      -webkit-box-shadow: 0 0 0 ${theme.spacings.small}
        ${theme.colors.lightGray} inset;
      filter: none;
      &::first-line {
        font-family: ${theme.font.family};
        font-size: ${theme.font.sizes.medium};
      }
    }
  `}
`
