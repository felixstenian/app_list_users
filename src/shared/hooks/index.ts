export { default as useKeyPress } from './useKeyPress'
export { default as useOnClickOutside } from './useOnClickOutside'
export { default as useSnackBar } from './useSnackBar'
