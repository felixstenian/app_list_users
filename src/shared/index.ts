export * as types from './types'
export * as helpers from './helpers'
export * as utils from './utils'
