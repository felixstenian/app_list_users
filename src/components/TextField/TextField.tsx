import { InputHTMLAttributes, useState } from 'react'
import * as Style from './styles'

export type TextFieldProps = {
  label?: string
  icon?: React.ReactNode
  iconPosition?: 'left' | 'right'
  placeholder?: string
  initialValue?: string
  onInputChange?: (value: string) => void
} & InputHTMLAttributes<HTMLInputElement>

const TextField = ({
  label,
  icon,
  placeholder,
  iconPosition = 'right',
  initialValue = '',
  onInputChange,
  name,
  ...props
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue)

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = e.currentTarget.value
    setValue(newValue)

    !!onInputChange && onInputChange(newValue)
  }

  return (
    <Style.Wrapper>
      {!!label && <Style.Label htmlFor={name}>{label}</Style.Label>}

      <Style.InputWraper>
        {!!icon && <Style.Icon iconPosition={iconPosition}>{icon}</Style.Icon>}
        <Style.Input
          type="text"
          onChange={onChange}
          value={value}
          iconPosition={iconPosition}
          name={name}
          placeholder={placeholder}
          {...(label ? { id: name } : {})}
          {...props}
        />
      </Style.InputWraper>
    </Style.Wrapper>
  )
}

export default TextField
