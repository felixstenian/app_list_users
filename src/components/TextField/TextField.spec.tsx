import { fireEvent, screen } from '@testing-library/react'

import { renderWithTheme } from 'shared/utils'
import TextField from '.'

describe('<TextField />', () => {
  it('1. should render the reading', () => {
    const { container } = renderWithTheme(<TextField label="TextField" />)

    // expect(screen.getByTestId('textfield')).toBeInTheDocument()

    expect(screen.getByText(/TextField/i)).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
  it('2. should render the label', () => {
    renderWithTheme(<TextField label="Name" name="name" />)
    expect(screen.getByText('Name')).toBeInTheDocument()
  })
  it('3. should render the icon', () => {
    const icon = <span>Icon</span>
    renderWithTheme(<TextField icon={icon} name="name" />)
    expect(screen.getByText('Icon')).toBeInTheDocument()
  })
  it('4. should render the placeholder', () => {
    renderWithTheme(<TextField placeholder="Name" name="name" />)
    expect(screen.getByPlaceholderText('Name')).toBeInTheDocument()
  })
  it('5. should call onInputChange when input value changes', () => {
    const onInputChange = jest.fn()
    renderWithTheme(
      <TextField
        name="name"
        onInputChange={onInputChange}
        data-testid="input"
      />
    )
    fireEvent.change(screen.getByTestId('input'), { target: { value: 'John' } })
    expect(onInputChange).toHaveBeenCalledWith('John')
  })
})
