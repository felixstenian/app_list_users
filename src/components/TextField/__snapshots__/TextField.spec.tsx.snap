// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> 1. should render the reading 1`] = `
.c0 {
  font-size: 1.4rem;
  color: #030517;
  cursor: pointer;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4rem;
  border: 0.2rem solid;
  background: #EAEAEA;
  padding: 0.8rem;
  border-color: #EAEAEA;
}

.c1:focus-within {
  box-shadow: 0 0 0.5rem #8F8F8F;
}

.c2 {
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  padding-right: 1.6rem;
  background: transparent;
  border: 0;
  outline: none;
  width: calc(100% - 2.2rem);
}

.c2:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 2.4rem #EAEAEA inset;
  -webkit-filter: none;
  filter: none;
}

.c2:-webkit-autofill::first-line {
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
}

<div
  class=""
>
  <label
    class="c0"
  >
    TextField
  </label>
  <div
    class="c1"
  >
    <input
      class="c2"
      type="text"
      value=""
    />
  </div>
</div>
`;
